core/backend.ml

```diff
@@ -66,22 +96,35 @@ struct
           print_bindings
         );
         IrCheck.Typecheck.bindings;
+        debug_tell "typechecked prelude IR";
       ]
 
 
-    let main_pipeline perform_optimisations = [
-        only_if perform_optimisations (measure "optimise" (run optimisation_pipeline));
+    let main_pipeline perform_optimisations () = [
+        only_if
+          perform_optimisations
+          (measure "optimise" (run optimisation_pipeline));
         Closures.program Lib.primitive_vars;
-        perform_for_side_effects (BuildTables.program Lib.primitive_vars);
-
-        only_if_set Basicsettings.Ir.typecheck_ir (run typechecking_pipeline);
+        perform_for_side_effects
+          (BuildTables.program Lib.primitive_vars);
+        only_if_any_set
+          [Basicsettings.Ir.typecheck_ir; Basicsettings.Ir.simplify_types]
+          (run simplify_type_structure_program);
+        only_if_set
+          Basicsettings.Ir.typecheck_ir
+          (run typechecking_pipeline);
       ]
 
-    let prelude_pipeline = [
+    let prelude_pipeline () = [
         (* May perform some optimisations here that are safe to do on the prelude *)
         (fun tenv globals -> Closures.bindings tenv Lib.primitive_vars globals);
         (fun tenv globals -> BuildTables.bindings tenv Lib.primitive_vars globals; globals);
-        only_if_set Basicsettings.Ir.typecheck_ir (run prelude_typechecking_pipeline);
+        only_if_any_set
+          [Basicsettings.Ir.typecheck_ir; Basicsettings.Ir.simplify_types]
```

Comment:
@frank-emrich, can you clarify why simplifying pass happens only when IR typechecking is enabled?